src/plan_expand_hypertable.c

```diff
@@ -306,12 +309,112 @@ get_chunk_oids(CollectQualCtx *ctx, PlannerInfo *root, RelOptInfo *rel, Hypertab
+			/* Exclude the root parent table */
+			return list_delete_oid(child_oids, ht->main_table_relid);
+		}
 	}
 	else
 		return get_explicit_chunk_oids(ctx, ht);
 }
 
+#if !(PG96 || PG10)
+
+/*
+ * Create partition expressions for a hypertable.
+ *
+ * Build an array of partition expressions where each element is a list of
+ * expression for a particular partitioning dimension.
+ *
+ * The partition expressions are used by, e.g., group_by_has_partkey() to check
+ * whether a GROUP BY clause covers all partitioning dimesions.
+ */
+static List **
+get_hypertable_partexprs(Hypertable *ht, Index varno)
+{
+	int i;
+	List **partexprs;
+
+	Assert(NULL != ht->space);
+
+	partexprs = palloc0(sizeof(List *) * ht->space->num_dimensions);
+
+	for (i = 0; i < ht->space->num_dimensions; i++)
+	{
+		Dimension *dim = &ht->space->dimensions[i];
+		Expr *expr;
+
+		/* Only functions on open dimensions need to be function expressions */
+		if (IS_OPEN_DIMENSION(dim) && NULL != dim->partitioning)
+			expr = (Expr *) dim->partitioning->partfunc.func_fmgr.fn_expr;
+		else
+		{
+			HeapTuple tuple = SearchSysCacheAttNum(ht->main_table_relid, dim->column_attno);
+			Form_pg_attribute att;
+
+			if (!HeapTupleIsValid(tuple))
+				elog(ERROR, "cache lookup failed for attribute");
+
+			att = (Form_pg_attribute) GETSTRUCT(tuple);
+
+			expr = (Expr *) makeVar(varno,
+									dim->column_attno,
+									att->atttypid,
+									att->atttypmod,
+									att->attcollation,
+									0);
+
+			ReleaseSysCache(tuple);
+		}
+
+		partexprs[i] = list_make1(expr);
+	}
+
+	return partexprs;
+}
+
+#define PARTITION_STRATEGY_MULTIDIM 'm'
+
+/*
+ * Partition info for hypertables.
+ *
+ * Build a "fake" partition scheme for a hypertable that makes the planner
+ * believe this is a PostgreSQL partitioned table for planning purposes. In
+ * particular, this will make the planner consider partitionwise aggregations
+ * when applicable.
+ *
+ * Partitionwise aggregation can either be FULL or PARTIAL. The former means
+ * that the aggregation can be performed independently on each partition (chunk)
+ * without a finalize step which is needed in PARTIAL. FULL requires that the
+ * GROUP BY clause contains all hypertable partitioning dimensions. This
+ * requirement is enforced by creating a partitioning scheme that covers
+ * multiple attributes, i.e., on per dimension. This works well since the
+ * "shallow" (one-level hierarchy) of a multi-dimensional hypertable is similar
+ * to a one-level partitioned PostgreSQL table where the partitioning key covers
+ * multiple attributes.
+ *
+ * Note that we use a partition scheme with a strategy that does not exist in
+ * PostgreSQL. This makes PostgreSQL raise errors when this partition scheme is
+ * used in places that require a valid partition scheme with a supported
+ * strategy.
+ */
+static void
+build_hypertable_partition_info(Hypertable *ht, RelOptInfo *hyper_rel)
+{
+	PartitionScheme part_scheme = palloc0(sizeof(PartitionSchemeData));
```

Comment:
Hmm i think this needs caching and reusing.

```
The planner maintains a list of canonical
partition schemes (distinct PartitionSchemeData objects) so that RelOptInfo of
any two partitioned relations with same partitioning scheme point to the same
PartitionSchemeData object.
```

Reply:
Yeah, I did look into this. My understanding is that this is only to avoid creating a lot of `PartitionScheme` objects that are duplicates. In any case, I don't think we can make use of that cache since (1) it is tied to the relcache and (2) the `PartitionScheme` object we create here is only partial so it is not a fully valid object to put in a cache. Unfortunately, we cannot create a fully valid `PartitionScheme`  since the way we partition is not compatible with PG11 partitioning. 

This is admittedly a bit of a hack, but not sure there's a way around it.

Reply:
the distinct PartitionSchema live in PlannerInfo::part_schemes, if we were to add our own PartitionSchema to that list and look in that list first before creating a new one this would enable partitionwise join which does pointer comparison on the PartitionSchema, but I guess we can skip that for now and keep it in mind for later.
https://insane.rocks/pg11/plancat_8c.html#a1e7cb48d0632cb52ad14d282600c0617

Reply:
Not strictly against it, but worried about adding a `PartitionScheme` to that list if that object is not fully valid. And, I am not sure we can make it fully valid since we aren't (at this point at least) compatible with PG11 partitioning.